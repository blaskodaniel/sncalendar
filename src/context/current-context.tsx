import { ConstantContent, ODataParams } from '@sensenet/client-core'
import { GenericContent } from '@sensenet/default-content-types'
import React, { useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { useRepository } from '../hooks/use-repository'
import { useInjector } from '../hooks/use-injector'

export const CurrentContentContext = React.createContext<GenericContent>(ConstantContent.PORTAL_ROOT)
export const CurrentContentProvider: React.FunctionComponent<{
  idOrPath: number | string
  onContentLoaded?: (content: GenericContent) => void
  oDataOptions?: ODataParams<GenericContent>
}> = props => {
  const [loadLock] = useState(new Semaphore(1))
  const [content, setContent] = useState<GenericContent>(ConstantContent.PORTAL_ROOT)
  const repo = useRepository()
  const injector = useInjector()
  const [reloadToken, setReloadToken] = useState(1)
  const reload = () => setReloadToken(Math.random())

  useEffect(() => {
    const events = injector.getEventHub(repo.configuration.repositoryUrl)
    const subscriptions = [
      events.onContentModified.subscribe((c: any) => {
        if (c.content.Id === content.Id) {
          reload()
        }
      }),
    ]
    return () => subscriptions.forEach(s => s.dispose())
  }, [repo, content, injector])

  const [error, setError] = useState<Error | undefined>()

  useEffect(() => {
    const ac = new AbortController()
    if (props.idOrPath) {
      ;(async () => {
        try {
          const response = await repo.load({
            idOrPath: props.idOrPath,
            requestInit: { signal: ac.signal },
            oDataOptions: props.oDataOptions,
          })
          setContent(response.d)
          props.onContentLoaded && props.onContentLoaded(response.d)
        } catch (err) {
          if (!ac.signal.aborted) {
            setError(err)
          }
        } finally {
          loadLock.release()
        }
      })()
    }
    return () => ac.abort()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [repo, props.idOrPath, reloadToken, loadLock])

  if (error) {
    throw error
  }

  return (
    <CurrentContentContext.Provider value={content as GenericContent}>{props.children}</CurrentContentContext.Provider>
  )
}
