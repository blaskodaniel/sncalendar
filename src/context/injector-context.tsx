import { Injector } from '@furystack/inject/dist/Injector'
import { ConsoleLogger } from '@furystack/logging'

import React from 'react'

export const snInjector = new Injector()
snInjector.options.owner = 'SnApp'
snInjector.useLogging(ConsoleLogger)

export const InjectorContext = React.createContext(snInjector)
