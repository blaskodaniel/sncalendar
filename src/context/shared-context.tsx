import React, { createContext, Dispatch, useState } from 'react'
import { GenericContent } from '@sensenet/default-content-types'
import { EditPropertiesDialog } from '../components/edit-dialog'
import CalendarEvent from '../CalendarEvent-type'
import { DialogComponent } from '../components/view-dialog'

export const SharedContext = createContext<{
  openeditmodal: boolean
  setOpeneditmodal: Dispatch<React.SetStateAction<boolean>>
  opendisplaymodal: boolean
  setOpendisplaymodal: Dispatch<React.SetStateAction<boolean>>
  setEvent: Dispatch<React.SetStateAction<GenericContent | undefined>>
}>(null as any)

const SharedProvider: React.FunctionComponent<any> = props => {
  const [openeditmodal, setOpeneditmodal] = useState(false)
  const [opendisplaymodal, setOpendisplaymodal] = useState(false)
  const [event, setEvent] = useState<CalendarEvent>(null as any)

  return (
    <>
      <SharedContext.Provider
        value={{ openeditmodal, setOpeneditmodal, setEvent, opendisplaymodal, setOpendisplaymodal }}>
        {props.children}
        {openeditmodal && event != null ? (
          <EditPropertiesDialog
            content={event}
            dialogProps={{
              open: openeditmodal,
              onClose: () => setOpeneditmodal(false),
              keepMounted: false,
            }}
          />
        ) : null}
        {opendisplaymodal && event != null ? (
          <DialogComponent open={opendisplaymodal} content={event} onClose={() => setOpendisplaymodal(false)} />
        ) : null}
      </SharedContext.Provider>
    </>
  )
}
export default SharedProvider
