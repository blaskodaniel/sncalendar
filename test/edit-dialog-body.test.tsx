import { mount, shallow } from 'enzyme'
import React from 'react'
import { Repository } from '@sensenet/client-core'
import { act } from 'react-dom/test-utils'
import EditPropertiesDialogBody from '../src/components/edit-dialog-body'
import { SharedContext } from '../src/context/shared-context'
import { RepositoryContext } from '../src/context/repository-provider'
import { MaterialDialogProps, mockContent } from './mocks/test-objects'

jest.mock('react-quill')

describe('EditPropertiesDialogBody', () => {
  const testprops = {
    contentId: 135121,
    dialogProps: MaterialDialogProps,
  }
  const content = mockContent
  const shareobject = {
    openeditmodal: true,
    setOpeneditmodal: jest.fn(),
    setEvent: jest.fn(),
    opendisplaymodal: false,
    setOpendisplaymodal: jest.fn(),
    event: content,
    refreshcalendar: true,
    setRefreshcalendar: jest.fn(),
    opennewmodal: true,
    setOpennewmodal: jest.fn(),
    setOpennoti: jest.fn(),
  }

  it('EditPropertiesDialogBody snapshot', () => {
    const l = shallow(<EditPropertiesDialogBody {...(testprops as any)} />)
    expect(l).toMatchSnapshot()
  })

  it('call onSubmit', async () => {
    const repo = new Repository()
    repo.load = function load() {
      return Promise.resolve({ d: content } as any)
    }
    repo.patch = function patch() {
      return Promise.resolve({ d: { ...content, Path: '/Root/Path/A' } } as any)
    }
    let wrapper: any

    await act(async () => {
      wrapper = mount(
        <RepositoryContext.Provider value={repo as any}>
          <SharedContext.Provider value={shareobject}>
            <EditPropertiesDialogBody {...(testprops as any)} />
          </SharedContext.Provider>
        </RepositoryContext.Provider>,
      )
    })

    const form = wrapper.update().find('form')
    await act(async () => {
      form.simulate('submit')
    })

    expect(shareobject.setEvent).toBeCalledWith({ ...content, Path: '/Root/Path/A' })
    expect(shareobject.setRefreshcalendar).toBeCalledWith(!shareobject.refreshcalendar)
  })
})
