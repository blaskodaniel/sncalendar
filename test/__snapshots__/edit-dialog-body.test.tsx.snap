// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPropertiesDialogBody EditPropertiesDialogBody snapshot 1`] = `
<Component
  idOrPath={135121}
  oDataOptions={
    Object {
      "select": "all",
    }
  }
>
  <ContextConsumer>
    <Component />
  </ContextConsumer>
</Component>
`;
